experiment/selinux/src/xattrs/xattr.rs

```diff
@@ -0,0 +1,105 @@
+    // SELinux label is not so big, so we allocate 1024 bytes for the buffer.
+    let mut buf = vec![0u8; 1024];
+    let size = rfs::getxattr(fpath, attr, &mut buf)
+        .map_err(|e| XattrError::GetXattr(e.to_string()))?;
+    let mut value = String::from_utf8_lossy(&buf[..size]).into_owned();
+    if (!value.is_empty()) && (value.ends_with('\x00')) {
+        value = value[..value.len() - 1].to_string();
+    }
+    Ok(value)
+}
+
+// function similar with lgetxattr in go-selinux repo.
+// lget_xattr returns the value of an extended attribute attr set for path.
+pub fn lget_xattr(fpath: &Path, attr: &str) -> Result<String, XattrError> {
+    // SELinux label is not so big, so we allocate 1024 bytes for the buffer.
+    let mut buf = vec![0u8; 1024];
+    let size = rfs::lgetxattr(fpath, attr, &mut buf)
+        .map_err(|e| XattrError::LGetXattr(e.to_string()))?;
+    let mut value = String::from_utf8_lossy(&buf[..size]).into_owned();
+    if (!value.is_empty()) && (value.ends_with('\x00')) {
+        value = value[..value.len() - 1].to_string();
+    }
+    Ok(value)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::xattrs::*;
+    use std::fs::{self, File};
+    use std::path::Path;
+    use std::io::Write;
+    use std::os::unix::fs::symlink;
+
+    fn create_temp_file(content: &[u8], file_name: &str) {
+        let path = Path::new(file_name);
+        let mut file = File::create(&path).expect("Failed to create file");
+        file.write_all(content).expect("Failed to write to file");
+        file.sync_all().expect("Failed to sync file");
+    }
+
+    #[test]
+    fn test_set_xattr_and_get_xattr() {
+        // Because of the permission issue, "selinux.security" can't be used here.
+        let attr_name = "user.test_attr";
+        let attr_value = "system_u:object_r:some_label_t";
+        let file_name = "test_file";
+        let file_path = Path::new(file_name);
+        create_temp_file(b"test", file_name);
+
+        set_xattr(&file_path, attr_name, attr_value.as_bytes()).expect("Failed to set xattr");
+        let actual = get_xattr(&file_path, attr_name).expect("Failed to get xattr");
+        assert_eq!(actual, attr_value);
+        fs::remove_file(file_name).expect("Failed to remove test file");
+    }
+
+    #[test]
+    fn test_lset_xattr_and_lget_xattr() {
+        // Because of the permission issue, "selinux.security" can't be used here.
+        let attr_name = "user.test_attr";
+        let attr_value = "system_u:object_r:some_label_t";
+        let file_name = "../test_file";
+        let file_path = Path::new(file_name);
+        let symlink_path = Path::new("test_symlink");
+        create_temp_file(b"test", file_name);
+        symlink(&file_path, &symlink_path).expect("Failed to create symlink");
+
+        lset_xattr(&symlink_path, attr_name, attr_value.as_bytes()).expect("Failed to lset xattr");
+        let actual = lget_xattr(&symlink_path, attr_name).expect("Failed to lget xattr");
```

Comment:
@utam0k @YJDoc2 
I wrote the test, but it failed on my codespace.
Do you know how to enable labeling lset_xattr on this environment ?

```
---- xattrs::xattr::tests::test_lset_xattr_and_lget_xattr stdout ----
thread 'xattrs::xattr::tests::test_lset_xattr_and_lget_xattr' panicked at src/xattrs/xattr.rs:99:69:
Failed to lset xattr: LSetXattr("Operation not permitted (os error 1)")
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Reply:
Hey, so after digging through https://www.man7.org/linux/man-pages/man7/xattr.7.html , https://www.man7.org/linux/man-pages/man2/listxattr.2.html, https://www.man7.org/linux/man-pages/man1/setfattr.1.html etc. I think the reason behind this is that `user.*` attributes are not supported for symlinks. Unfortunately the rest (system,security and trusted) all require CAP_SYS_ADMIN i.e. root perms to be set, which means we cannot really test this in a straightforward way as far as I can think. Hope this helps.

I also saw the code in rustix itself, and even they don't seem to be testing actual get/set for symlinks, only some error conditions.

Reply:
I see, thank you so much for your investigation.
Though I will investigate a bit more on whether there is a way running a test for this, I think we can skip lset_xattr test here.

Reply:
Added "ignore"